scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketSource.java

```diff
@@ -578,22 +615,60 @@ public String toString() {
     @SuppressWarnings("unchecked")
     private void writeObject(ObjectOutputStream outStream) throws IOException {
       SerializableCoder.of(TupleTag.class).encode(tupleTag, outStream);
-      ListCoder.of(StringUtf8Coder.of()).encode(inputDirectories, outStream);
-      outStream.writeUTF(filenameSuffix);
-      outStream.writeObject(fileOperations);
       outStream.writeObject(predicate);
       outStream.writeObject(keying);
+
+      long numDistinctFileSuffixes =
+          directories.values().stream().map(KV::getKey).distinct().count();
+      long numDistinctFileOperations =
+          directories.values().stream().map(kv -> kv.getValue().getClass()).distinct().count();
+
+      // If all partitions use the same file operations type, don't keep re-encoding it
```

Comment:
In the past serialized transform size has been an issue (if hundreds+ of partitions are being read), but this solution is definitely a bit convoluted 😅

Reply:
You could potentially add a header to write out the distinct file operations with indicies and have a reference to those indicies in each of the output objects. Slight increase in overhead if all file ops are different, but would decrease payload size even more if there are some (but not exactly 1) shared operations

Reply:
that's a good idea - will give it a try